test/common/stats/udp_statsd_test.cc

```diff
@@ -29,21 +30,73 @@ TEST_P(UdpStatsdSinkTest, InitWithIpAddress) {
   Network::Address::InstanceConstSharedPtr server_address =
       Network::Utility::parseInternetAddressAndPort(
           fmt::format("{}:8125", Network::Test::getLoopbackAddressUrlString(GetParam())));
-  UdpStatsdSink sink(tls_, server_address);
+  UdpStatsdSink sink(tls_, server_address, false);
   int fd = sink.getFdForTests();
   EXPECT_NE(fd, -1);
 
   // Check that fd has not changed.
+  std::vector<Tag> tags;
   NiceMock<MockCounter> counter;
   counter.name_ = "test_counter";
+  ON_CALL(counter, tags()).WillByDefault(ReturnRef(tags));
   sink.flushCounter(counter, 1);
 
   NiceMock<MockGauge> gauge;
   gauge.name_ = "test_gauge";
+  ON_CALL(gauge, tags()).WillByDefault(ReturnRef(tags));
   sink.flushGauge(gauge, 1);
 
   NiceMock<MockHistogram> timer;
   timer.name_ = "test_timer";
+  ON_CALL(timer, tags()).WillByDefault(ReturnRef(tags));
+  sink.onHistogramComplete(timer, 5);
+
+  EXPECT_EQ(fd, sink.getFdForTests());
+
+  if (GetParam() == Network::Address::IpVersion::v4) {
+    EXPECT_EQ("127.0.0.1:8125", Network::Address::peerAddressFromFd(fd)->asString());
+  } else {
+    EXPECT_EQ("[::1]:8125", Network::Address::peerAddressFromFd(fd)->asString());
+  }
+  tls_.shutdownThread();
+}
+
+class UdpStatsdSinkWithTagsTest : public testing::TestWithParam<Network::Address::IpVersion> {};
+INSTANTIATE_TEST_CASE_P(IpVersions, UdpStatsdSinkWithTagsTest,
+                        testing::ValuesIn(TestEnvironment::getIpVersionsForTest()));
+
+TEST_P(UdpStatsdSinkWithTagsTest, InitWithIpAddress) {
+  NiceMock<ThreadLocal::MockInstance> tls_;
+  // UDP statsd server address.
+  Network::Address::InstanceConstSharedPtr server_address =
+      Network::Utility::parseInternetAddressAndPort(
+          fmt::format("{}:8125", Network::Test::getLoopbackAddressUrlString(GetParam())));
+  UdpStatsdSink sink(tls_, server_address, true);
+  int fd = sink.getFdForTests();
+  EXPECT_NE(fd, -1);
+
+  // Check that fd has not changed.
+  std::vector<Tag> tags = {Tag{"node", "test"}};
+  std::string name = "test_counter";
+  NiceMock<MockCounter> counter;
+  counter.name_ = name;
+  ON_CALL(counter, tagExtractedName()).WillByDefault(ReturnRef(name));
+  ON_CALL(counter, tags()).WillByDefault(ReturnRef(tags));
+
+  sink.flushCounter(counter, 1);
```

Comment:
I don't see the actual expected statsd raw output in any of these tests? Shouldn't we have something like this?

I think this is an issue with the existing tests; there's no validation that we're actually building the correct string.

Reply:
Yes, agree to add the validation test. I'll try it in this PR.

Reply:
Thanks, this would be a nice improvement to the existing tests.

Reply:
It was a little bit difficult for me, but I tried with changing `Writer`'s responsibility. 
How about this one: e3487df436fc02087cb2193d9bbb86e7381ae742